src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -952,43 +1127,72 @@ Containers::String UfbxImporter::doMaterialName(UnsignedInt id) {
+
     struct MaterialMappingList {
         Containers::ArrayView<const MaterialMapping> mappings;
         Containers::ArrayView<ufbx_material_map> maps;
     };
-
     MaterialMappingList mappingLists[] = {
-        { Containers::arrayView(materialMapsFbx), Containers::arrayView(material->fbx.maps) },
-        { Containers::arrayView(materialMapsPbr), Containers::arrayView(material->pbr.maps) },
+        { Containers::arrayView(materialMappingPbr), Containers::arrayView(material->pbr.maps) },
+        { Containers::arrayView(materialMappingFbx), Containers::arrayView(material->fbx.maps) },
     };
 
-    Containers::BitArray addedAttributes { ValueInit, 128 };
+    /* Flexible for custom layers (layered textures) */
+    Containers::Array<Containers::Array<MaterialAttributeData>> attributeLayers{UfbxMaterialLayerCount};
 
-    UnsignedInt maxAttributes = 0;
-    for (MaterialMappingList &list : mappingLists) {
-        /* Each mapping may generate multiple mappings, estimate for wrost case
-           Value + Texture + TextureMatrix + TextureCoordinates */
-        maxAttributes += list.mappings.size() * 4;
+    Containers::String attribute;
 
-        #ifndef CORRADE_NO_ASSERT
-        for (const MaterialMapping &mapping : list.mappings) {
-            CORRADE_INTERNAL_ASSERT(UnsignedInt(mapping.attrib) < addedAttributes.size());
-            CORRADE_INTERNAL_ASSERT(UnsignedInt(mapping.texture) < addedAttributes.size());
-            CORRADE_INTERNAL_ASSERT(UnsignedInt(mapping.textureCoordinates) < addedAttributes.size());
-            CORRADE_INTERNAL_ASSERT(UnsignedInt(mapping.textureMatrix) < addedAttributes.size());
+    MaterialTypes types;
+
+    /* Do some feature detection */
+
+    /* If we have DiffuseColor specified from the FBX properties the fallback
+       FBX material should be quite well defined. */
+    if (material->fbx.diffuse_color.has_value) {
+        types |= MaterialType::Phong;
+    }
+
+    /* ufbx supports glossiness through ufbx_material_map::texture_inverted and
+       inverts the values internally. This is done due to 3ds Max having material
+       models where the roughness inversion is controlled via a property
+       "roughness_inv", "coat_roughness_inv", etc. But seems to be actually very
+       confusing to users. Maybe there should be just glossiness maps instead.
+       If this change is done need to clean up some code here.. */
+    bool roughnessIsGlossiness = material->pbr.roughness.texture_inverted;
+    if (material->features.pbr.enabled) {
+        if (material->features.metalness.enabled && !roughnessIsGlossiness) {
+            types |= MaterialType::PbrMetallicRoughness;
+        } else if (material->features.specular.enabled && roughnessIsGlossiness) {
+            types |= MaterialType::PbrSpecularGlossiness;
         }
-        #endif
+        /* Missing: PbrMetallicGlossiness, PbrSpecularRoughness, but these
+           are quite rare in practice */
```

Comment:
I briefly thought about having both `PbrMetallicRoughness` and `PbrSpecularGlossiness` set, but that would mean neither of them is a complete representation, which would only lead to misery. I guess best is to just not set any of the two in that case, as you do.

IIRC the `KHR_materials_specular` extension made it possible to convert specular/glossiness losslessly(?) to the regular metallic/roughness, and so the [KHR_materials_pbrSpecularGlossiness](https://github.com/KhronosGroup/glTF/tree/main/extensions/2.0/Archived/KHR_materials_pbrSpecularGlossiness) extension got archived because it's now redundant. So I imagine I could have a tool that converts those unruly metallic/glossiness etc. FBX materials to metallic/roughness+specular, together with inverting texture data and whatnot. The importer should just output the material attributes as vanilla as possible in my opinion.